### PR TITLE
Support conditional children in most JSX components

### LINKED ISSUE #2505: Support conditional rendering with booleans
This is currently not supported, we should support it:

```
<Box>
  {condition && <Text>Foo</Text>}
</Box>
```

### PR DESCRIPTION
**Pull request by Mrtenz**
This adds support for conditional children in most JSX components. The exceptions are `Field`, `Row`, and other components which don't have children in the first place. For example, the following is now possible:

```tsx
<Box direction="horizontal" alignment="space-between">
  {condition && <Text>Hello</Text>}
</Box>
```

To support this, I've added `boolean` as valid prop for the JSX components. The `getJsxChildren` function filters out any falsy or literal `true` values.

I've also added a small contributing guide for conventions around JSX components, including the support for conditional children.

Closes #2505.

### CODE REVIEW

File: packages/snaps-sdk/src/jsx/validation.ts [hunk 1]
```diff
@@ -161,7 +161,7 @@ export const OptionStruct: Describe<OptionElement> = element('Option', {
 export const DropdownStruct: Describe<DropdownElement> = element('Dropdown', {
   name: string(),
   value: optional(string()),
-  children: maybeArray(OptionStruct),
+  children: maybeArray(nullUnion([OptionStruct, boolean()])),
```
Comment (FrederikBolding) [inline packages/snaps-sdk/src/jsx/validation.ts:164]: Realizing that this means that null isn't allowed in Dropdowns currently.

Would it be easier to have a `jsxChildren` struct that takes a list of structs and combines that with `boolean()`, `null` etc?

File: packages/snaps-sdk/src/jsx/components/Link.ts [hunk 2]
```diff
@@ -6,7 +6,7 @@ import type { StandardFormattingElement } from './formatting';
  * The children of the {@link Link} component.
  */
 export type LinkChildren = MaybeArray<
```
Comment (FrederikBolding) [inline packages/snaps-sdk/src/jsx/components/Link.ts:8]: Same question with regards to types, would it be easier to have a shared type that adds `boolean | null` and whatever else

File: packages/snaps-sdk/src/jsx/validation.ts (hunk 1 above)
Comment (Mrtenz) [reply packages/snaps-sdk/src/jsx/validation.ts:164]: Sure, that does sound easier.

File: packages/snaps-sdk/src/jsx/components/form/Field.ts [hunk 3]
```diff
@@ -20,7 +20,8 @@ export type FieldProps = {
     | DropdownElement
     | FileInputElement
     | InputElement
-    | CheckboxElement;
+    | CheckboxElement
+    | boolean;
```
Comment (FrederikBolding) [inline packages/snaps-sdk/src/jsx/components/form/Field.ts:24]: An empty Field doesn't make much sense, perhaps worth it to skip

File: packages/snaps-sdk/CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+ *
+ * @property name - The name of the dropdown. This is used to identify the
+ * state in the form data.
+ * @property value - The selected value of the dropdown.
+ * @property children - The children of the dropdown.
+ */
+export type DropdownProps = {
+  name: string;
+  value?: string | undefined;
+  children: MaybeArray<OptionElement | boolean>;
```
Comment (FrederikBolding) [inline packages/snaps-sdk/CONTRIBUTING.md:46]: `MaybeArray` doesn't exist anymore, not sure if it needs to?

File: packages/snaps-sdk/CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+#### Children
+
+In most cases, the children prop should be defined as `SnapsChildren` to allow
+for both a single child element or an array of child elements. There are some
+exceptions to this rule, such as when the component only accepts a single child
+element (e.g., `Field`).
+
+Children should also accept `boolean` and `null` values to allow for conditional
+rendering of child elements.
```
Comment (FrederikBolding) [inline packages/snaps-sdk/CONTRIBUTING.md:81]:
```suggestion
rendering of child elements. The `SnapsChildren` type handles this edge-case automatically.
```

File: packages/snaps-sdk/src/jsx/components/form/Option.ts [hunk 6]
```diff
@@ -9,7 +12,7 @@ import { createSnapComponent } from '../../component';
  */
 type OptionProps = {
   value: string;
-  children: string;
+  children: StringElement;
```
Comment (FrederikBolding) [inline packages/snaps-sdk/src/jsx/components/form/Option.ts:15]: Hmm, wondering if we want to support this 🤔

Comment (Mrtenz) [reply packages/snaps-sdk/src/jsx/components/form/Option.ts:15]: Will revert

File: packages/snaps-sdk/src/jsx/validation.ts [hunk 7]
```diff
@@ -263,10 +266,10 @@ export const AddressStruct: Describe<AddressElement> = element('Address', {
  * A struct for the {@link BoxElement} type.
  */
 export const BoxStruct: Describe<BoxElement> = element('Box', {
-  children: maybeArray(
+  children: children(
     // eslint-disable-next-line @typescript-eslint/no-use-before-define
-    nullable(lazy(() => BoxChildStruct)),
-  ) as unknown as Struct<MaybeArray<GenericSnapElement | null>, null>,
+    [lazy(() => BoxChildStruct), boolean()],
```
Comment (FrederikBolding) [inline packages/snaps-sdk/src/jsx/validation.ts:271]: Can't we omit `boolean()` here?

Comment (Mrtenz) [reply packages/snaps-sdk/src/jsx/validation.ts]: Good catch, forgot to update this after simplifying the types/validation with `children`.

File: packages/snaps-sdk/src/jsx/validation.ts [hunk 8]
```diff
@@ -206,45 +216,39 @@ export const FieldStruct: Describe<FieldElement> = element('Field', {
...
-        // eslint-disable-next-line @typescript-eslint/no-use-before-define
-        lazy(() => ItalicStruct) as unknown as Struct<
-          SnapElement<JsonObject, 'Italic'>
-        >,
-      ]),
-    ),
-  ),
+  children: children([
+    string(),
+    boolean(),
```
Comment (FrederikBolding) [inline packages/snaps-sdk/src/jsx/validation.ts:232]:
```suggestion
```

File: packages/snaps-sdk/src/jsx/validation.ts [hunk 9]
```diff
@@ -206,45 +216,39 @@ export const FieldStruct: Describe<FieldElement> = element('Field', {
...
-        // eslint-disable-next-line @typescript-eslint/no-use-before-define
-        lazy(() => BoldStruct) as unknown as Struct<
-          SnapElement<JsonObject, 'Bold'>
-        >,
-      ]),
-    ),
-  ),
+  children: children([
+    string(),
+    boolean(),
```
Comment (FrederikBolding) [inline packages/snaps-sdk/src/jsx/validation.ts:246]:
```suggestion
```

File: packages/snaps-sdk/src/jsx/validation.ts [hunk 10]
```diff
@@ -320,16 +324,20 @@ export const ImageStruct: Describe<ImageElement> = element('Image', {
  */
 export const LinkStruct: Describe<LinkElement> = element('Link', {
   href: string(),
-  children: maybeArray(nullable(nullUnion([FormattingStruct, string()]))),
+  children: children([FormattingStruct, string(), boolean()]),
```
Comment (FrederikBolding) [inline packages/snaps-sdk/src/jsx/validation.ts:327]:
```suggestion
  children: children([FormattingStruct, string()]),
```

File: packages/snaps-sdk/src/jsx/validation.ts [hunk 11]
```diff
@@ -320,16 +324,20 @@ export const ImageStruct: Describe<ImageElement> = element('Image', {
...
 /**
  * A struct for the {@link TextElement} type.
  */
 export const TextStruct: Describe<TextElement> = element('Text', {
-  children: maybeArray(
-    nullable(nullUnion([string(), BoldStruct, ItalicStruct, LinkStruct])),
-  ),
+  children: children([
+    string(),
+    boolean(),
```
Comment (FrederikBolding) [inline packages/snaps-sdk/src/jsx/validation.ts:336]:
```suggestion
```

File: packages/snaps-sdk/src/jsx/validation.ts [hunk 12]
```diff
@@ -345,6 +353,7 @@ export const TooltipChildStruct = nullUnion([
   ItalicStruct,
   LinkStruct,
   ImageStruct,
+  boolean(),
```
Comment (FrederikBolding) [inline packages/snaps-sdk/src/jsx/validation.ts:356]:
```suggestion
```

**Review by FrederikBolding — APPROVED** [review]

**Pull request merged**